### PR TITLE
Logical short circuit

### PR DESCRIPTION
**Pull request by eellison**
Adding short circuit evaluation to AND or OR. The second expression of and AND or OR gets lifted into an if branch, which is conditionally evaluated. 

BatchOps was using the expression `dims = dims1 or dims2`, where dims is often an empty tensor. This nows throws an error, because dims1 gets cast to a boolean, and you can't convert an empty tensor to a scalar. It now matches the behavior of pytorch in python. 

One thing that came up is if the second expression in an and/or in python gets returned, it does not get coerced to a boolean. 

`tensor == (False or tensor)`
`tensor == (True and tensor)`

We do not currently support this. 

edit: wording

### CODE REVIEW

File: torch/jit/batchop.py [hunk 1]
```diff
@@ -32,16 +32,22 @@ def batch_neg(data, mask, dims):
...
+        dims = dims2
+    return dims
 
 @torch.jit.script
 def batch_add(data1, mask1, dims1, data2, mask2, dims2, alpha_):
     alpha = float(alpha_)
     data = torch.add(data1, data2, alpha=alpha)
     mask = mask1 * mask2
-    dims = dims1 or dims2
+    dims = batch_nonempty_dims(dims1, dims2)
```
Comment (zdevito) [inline torch/jit/batchop.py]: These are bitwise ors: `dims1.__or__(dims2)`

File: test/test_jit.py [hunk 2]
```diff
@@ -3102,6 +3102,32 @@ def func(x, y):
...
+        def throwsAnd(t):
+            c0 = True and t[1]
+            print(c0)
+
+        t = torch.randn(0)
+        self.assertEqual(0, testNoThrows(torch.randn(0)))
+        self.assertExpectedGraph(testNoThrows.graph)
+        with self.assertRaises(RuntimeError):
+            throwsOr(t)
+        with self.assertRaises(RuntimeError):
```
Comment (apaszke) [inline test/test_jit.py]: Can we check the error message too?

File: torch/csrc/jit/script/compiler.cpp [hunk 3]
```diff
@@ -908,29 +908,61 @@ struct to_ir {
 
   Value* emitTernaryIf(const TernaryIf& expr) {
     Value* cond_value = emitCond(expr.cond());
+    auto first_expr = [&] {
+      return emitExpr(expr.true_expr());
+    };
+    auto second_expr = [&] {
```
Comment (apaszke) [inline torch/csrc/jit/script/compiler.cpp]: nit: `true_expr` and `false_expr` are much better names than `first` and `second`

**Review by apaszke — APPROVED** [review]

File: torch/csrc/jit/script/compiler.cpp (hunk 3 above)
Comment (eellison) [reply torch/csrc/jit/script/compiler.cpp]: agreed

File: test/test_jit.py (hunk 2 above)
Comment (eellison) [reply test/test_jit.py]: Yep will do!

**Pull request closed without merging**